src/main.rs

```diff
@@ -1,10 +1,66 @@
+use std::collections::HashSet;
 use std::fs;
 
-#[derive(Copy, Clone, Debug)]
+#[derive(Clone, Debug)]
 enum ByteSet {
-    Any,
-    Is(u8),
-    Not(u8),
+    Includes(HashSet<u8>),
+    Excludes(HashSet<u8>),
+}
+
+impl ByteSet {
+    pub fn any() -> ByteSet {
+        ByteSet::Excludes(HashSet::new())
+    }
+
+    pub fn is(b: u8) -> ByteSet {
+        ByteSet::Includes(HashSet::from([b]))
+    }
+
+    pub fn not(b: u8) -> ByteSet {
+        ByteSet::Excludes(HashSet::from([b]))
+    }
+
+    pub fn contains(&self, b: u8) -> bool {
+        match self {
+            ByteSet::Includes(included) => included.contains(&b),
+            ByteSet::Excludes(excluded) => !excluded.contains(&b),
+        }
+    }
+
+    pub fn is_disjoint(bs0: &ByteSet, bs1: &ByteSet) -> bool {
+        match (bs0, bs1) {
+            // Easy: check that the sets of contained bytes are disjoint
+            (ByteSet::Includes(included0), ByteSet::Includes(included1)) => {
+                HashSet::is_disjoint(included0, included1)
+            }
+            // If the set of contained bytes are a subset of the excluded bytes,
+            // then the byte sets are disjoint
+            (ByteSet::Includes(included), ByteSet::Excludes(excluded))
+            | (ByteSet::Excludes(excluded), ByteSet::Includes(included)) => {
+                HashSet::is_subset(included, excluded)
+            }
+            // Too hard to enumerate all of these: just say that they aren’t disjoint
+            (ByteSet::Excludes(_), ByteSet::Excludes(_)) => false,
```

Comment:
Is there any reason not to check [0..255] ? Seems like it might be a one liner :thinking:

Reply:
Not sure if I got it right!